File: 28. Find the Index of the First Occurrence in a String/main.py
class Solution:
    def strStr(self, haystack: str, needle: str) -> int:
        return haystack.find(needle)


class Solution:
    def strStr(self, haystack: str, needle: str) -> int:
        if needle in haystack:
            return haystack.index(needle)

        return -1


class Solution:
    def strStr(self, haystack: str, needle: str) -> int:
        try:
            return haystack.index(needle)
        except:
            return -1


class Solution:
    def strStr(self, haystack: str, needle: str) -> int:
        len1 = len(haystack)
        len2 = len(needle)

        for i in range(0, len1):
            j = 0
            while j < len2 and i + j < len1 and haystack[i + j] == needle[j]:
                j += 1

            if j == len2:
                return i

        return -1


class Solution:
    def strStr(self, haystack: str, needle: str) -> int:
        len1 = len(haystack)
        len2 = len(needle)

        if len1 < len2:
            return -1

        for i in range(0, len1 + 1 - len2):
            if haystack[i : i + len2] == needle:
                return i

        return -1

File: 28. Find the Index of the First Occurrence in a String/main.rs
impl Solution {
    pub fn str_str(haystack: String, needle: String) -> i32 {
        if let Some(pos) = haystack.find(&needle) {
            return pos as i32;
        }

        return -1;
    }
}

impl Solution {
    pub fn str_str(haystack: String, needle: String) -> i32 {
        let v: Vec<_> = haystack.match_indices(&needle).collect();
        if v.is_empty() {
            return -1;
        }

        return v[0].0 as i32;
    }
}

impl Solution {
    pub fn str_str(haystack: String, needle: String) -> i32 {
        let len1 = haystack.len();
        let len2 = needle.len();

        if len1 < len2 {
            return -1;
        }

        for i in 0..=len1 - len2 {
            if haystack[i..i + len2] == needle[..] {
                return i as i32;
            }
        }

        return -1;
    }
}

impl Solution {
    pub fn str_str(haystack: String, needle: String) -> i32 {
        let len1 = haystack.len();
        let len2 = needle.len();

        if len1 < len2 {
            return -1;
        }

        for i in 0..=len1 - len2 {
            if haystack.get(i..i + len2) == Some(&needle) {
                return i as i32;
            }
        }

        return -1;
    }
}

impl Solution {
    pub fn str_str(haystack: String, needle: String) -> i32 {
        let s1 = haystack.chars().collect::<Vec<_>>();
        let s2 = needle.chars().collect::<Vec<_>>();
        let len1 = s1.len();
        let len2 = s2.len();

        if len1 < len2 {
            return -1;
        }

        for i in 0..=len1 - len2 {
            let mut j = 0;
            while j < len2 && s1[i + j] == s2[j] {
                j += 1;
            }

            if j == len2 {
                return i as i32;
            }
        }

        return -1;
    }
}
